README.md
# ebpf_exporter

ebpf_exporter is an experimental Prometheus exporter which uses eBPF kprobes to efficiently record
a histogram of Linux bio request latencies and sizes.

The included BPF program is loosely based on the examples shipped with [IO Visor's BPF Compiler
Collection](https://github.com/iovisor/bcc), specifically the bitehist and disksnoop examples.

## Sample Output

Linux bio request latencies for each block device are recorded in log2 buckets separately for each
request operation type (read, write, flush, discard, etc), in microseconds. This should cover use
cases ranging from high speed flash-based devices, to legacy HDD devices.

```
# HELP ebpf_bio_req_latency A histogram of bio request latencies in microseconds.
# TYPE ebpf_bio_req_latency histogram
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="1"} 0
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="2"} 0
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="4"} 0
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="8"} 0
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="16"} 0
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="32"} 0
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="64"} 40
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="128"} 219
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="256"} 335
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="512"} 363
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="1024"} 428
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="2048"} 1649
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="4096"} 3498
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="8192"} 3614
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="16384"} 3760
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="32768"} 3777
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="65536"} 3780
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="131072"} 3780
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="262144"} 3780
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="524288"} 3780
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="1.048576e+06"} 3780
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="2.097152e+06"} 3780
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="4.194304e+06"} 3780
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="8.388608e+06"} 3780
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="1.6777216e+07"} 3780
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="3.3554432e+07"} 3780
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="6.7108864e+07"} 3780
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="1.34217728e+08"} 3780
ebpf_bio_req_latency_bucket{device="sda",operation="write",le="+Inf"} 3780
ebpf_bio_req_latency_sum{device="sda",operation="write"} 1.4306176e+07
ebpf_bio_req_latency_count{device="sda",operation="write"} 3780
```

Request sizes (in KiB) are also recorded in log2 buckets for each device:

```
# HELP ebpf_bio_req_size A histogram of bio request sizes in KiB.
# TYPE ebpf_bio_req_size histogram
ebpf_bio_req_size_bucket{device="sda",operation="write",le="1"} 8
ebpf_bio_req_size_bucket{device="sda",operation="write",le="2"} 8
ebpf_bio_req_size_bucket{device="sda",operation="write",le="4"} 2450
ebpf_bio_req_size_bucket{device="sda",operation="write",le="8"} 3050
ebpf_bio_req_size_bucket{device="sda",operation="write",le="16"} 3347
ebpf_bio_req_size_bucket{device="sda",operation="write",le="32"} 3602
ebpf_bio_req_size_bucket{device="sda",operation="write",le="64"} 3698
ebpf_bio_req_size_bucket{device="sda",operation="write",le="128"} 3714
ebpf_bio_req_size_bucket{device="sda",operation="write",le="256"} 3732
ebpf_bio_req_size_bucket{device="sda",operation="write",le="512"} 3777
ebpf_bio_req_size_bucket{device="sda",operation="write",le="1024"} 3780
ebpf_bio_req_size_bucket{device="sda",operation="write",le="2048"} 3780
ebpf_bio_req_size_bucket{device="sda",operation="write",le="4096"} 3780
ebpf_bio_req_size_bucket{device="sda",operation="write",le="8192"} 3780
ebpf_bio_req_size_bucket{device="sda",operation="write",le="16384"} 3780
ebpf_bio_req_size_bucket{device="sda",operation="write",le="32768"} 3780
ebpf_bio_req_size_bucket{device="sda",operation="write",le="+Inf"} 3780
ebpf_bio_req_size_sum{device="sda",operation="write"} 66400
ebpf_bio_req_size_count{device="sda",operation="write"} 3780
```

Note that histograms will only be exposed for devices that have actually performed IO since the BPF
program was loaded. If a device is only performing writes, no histogram will be present for reads,
and vice versa.

## Grafana Panel Samples

Grafana 5.1 and later supports Prometheus histograms, either as bar / line / point graphs, or as
heatmaps. See Grafana's [heatmap panel](http://docs.grafana.org/features/panels/heatmap/) guide for
more information.

![IO request latency histogram](img/io-req-latency-hist.png)

![IO request latency heatmap](img/io-req-latency-heatmap.png)


debian/changelog
prometheus-ebpf-exporter (0.0.1) UNRELEASED; urgency=low

  * Initial packaged version

 -- Daniel Swarbrick <dswarbrick@users.noreply.github.com>  Wed, 13 Jun 2018 16:52:34 +0200


bpf.go
// ebpf_exporter - A Prometheus exporter for Linux block IO statistics.
//
// Copyright 2018 Daniel Swarbrick
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

const bpfSource string = `
#include <uapi/linux/ptrace.h>
#include <linux/blkdev.h>
#include <linux/blk_types.h>

typedef struct disk_key {
	char disk[DISK_NAME_LEN];	// 32 bytes
	u8 req_op;
	u64 slot;
} disk_key_t;				// 48 bytes, with padding

const u8 max_io_lat_slot = 28;		// log2 range 1 us to ~2 mins
const u8 max_io_req_sz_slot = 16;	// log2 range 1 KiB to 32 MiB

// Hash to temporily hold the start time of each bio request - macro for
// BPF_TABLE("hash", _key_type, u64, _name, 10240). Increase if you expect
// more than 10K IO requests in flight.
BPF_HASH(start, struct request *);

// Histograms to hold IO request latency / size bucket values - macro for
// BPF_TABLE("histogram", _key_type, u64, _name, _size). Total number of
// buckets are shared amongst all devices and all request operation types.
// Unlike Prometheus histograms, these are sparse, so will only use a bucket
// if required. Since most request operations will be read or write, a good
// rule of thumb is: num_devices * 2 req_op types * 20 buckets each. Bear in
// mind that the amount of memory used will be (sizeof(_key_type) +
// sizeof(u64)) * _size, so the following will use 560 KiB each.
BPF_HISTOGRAM(io_lat, disk_key_t, 10240);
BPF_HISTOGRAM(io_req_sz, disk_key_t, 10240);

// Record start time of a request
int trace_req_start(struct pt_regs *ctx, struct request *req)
{
	u64 ts = bpf_ktime_get_ns();
	start.update(&req, &ts);
	return 0;
}

// Calculate request duration and store in appropriate histogram bucket
int trace_req_completion(struct pt_regs *ctx, struct request *req, unsigned int bytes)
{
	u64 *tsp, delta, slot;
	u8 req_op;

	// Fetch timestamp and calculate delta
	tsp = start.lookup(&req);
	if (tsp == 0) {
		return 0;   // missed issue
	}

	// Request duration, in microseconds
	delta = (bpf_ktime_get_ns() - *tsp) / 1000;

	// Request operation, e.g. REQ_OP_READ, REQ_OP_WRITE, etc.
	req_op = req->cmd_flags & REQ_OP_MASK;

	// Latency histogram key
	slot = bpf_log2l(delta);
	if (slot >= max_io_lat_slot)
		slot = max_io_lat_slot - 1;
	disk_key_t lat_key = {.slot = slot, .req_op = req_op};
	bpf_probe_read(&lat_key.disk, sizeof(lat_key.disk), req->rq_disk->disk_name);

	// Request size histogram key
	slot = bpf_log2(bytes / 1024);
	if (slot >= max_io_req_sz_slot)
		slot = max_io_req_sz_slot - 1;
	disk_key_t req_sz_key = {.slot = slot, .req_op = req_op};
	bpf_probe_read(&req_sz_key.disk, sizeof(req_sz_key.disk), req->rq_disk->disk_name);

	io_lat.increment(lat_key);
	io_req_sz.increment(req_sz_key);

	start.delete(&req);
	return 0;
}
`


main.go
// +build linux

// ebpf_exporter - A Prometheus exporter for Linux block IO statistics.
//
// Copyright 2018 Daniel Swarbrick
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"fmt"
	"net/http"
	"os"

	"github.com/go-kit/kit/log/level"
	"github.com/iovisor/gobpf/bcc"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/prometheus/common/promlog"
	"github.com/prometheus/common/promlog/flag"
	"github.com/prometheus/common/version"
	"gopkg.in/alecthomas/kingpin.v2"
)

const namespace = "ebpf"

var (
	listenAddress = kingpin.Flag("web.listen-address", "The address to listen on for HTTP requests.").Default(":9123").String()
)

func main() {
	allowedLevel := promlog.AllowedLevel{}
	flag.AddFlags(kingpin.CommandLine, &allowedLevel)
	kingpin.Version(version.Print("ebpf_exporter"))
	kingpin.HelpFlag.Short('h')
	kingpin.Parse()

	logger := promlog.New(allowedLevel)

	level.Info(logger).Log("msg", "Starting ebpf_exporter", "version", version.Info())
	level.Info(logger).Log("msg", "Build context", version.BuildContext())

	// Compile BPF code and return new module
	m := bcc.NewModule(bpfSource, []string{})
	defer m.Close()

	// Map of kprobe names from our BPF program to kernel function names, to which to attach.
	kprobes := map[string]string{
		"trace_req_start":      "blk_account_io_start",
		"trace_req_completion": "blk_account_io_completion",
	}

	// Load kprobes and attach them to kernel functions
	for kpName, fnName := range kprobes {
		if kp, err := m.LoadKprobe(kpName); err == nil {
			if err := m.AttachKprobe(fnName, kp); err != nil {
				fmt.Fprintf(os.Stderr, "Failed to attach %q to %q: %s\n", kpName, fnName, err)
				os.Exit(1)
			}
		} else {
			fmt.Fprintf(os.Stderr, "Failed to load %q: %s\n", kpName, err)
			os.Exit(1)
		}
	}

	prometheus.MustRegister(newExporter(m))

	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/html")
		w.Write([]byte(`<html>
	<head>
	<title>eBPF Exporter</title>
	<style>html { font-family: sans-serif; }</style>
	</head>
	<body>
	<h1>eBPF Exporter</h1>
	<p><a href="/metrics">Metrics</a></p>
	</body>
</html>`))
	})
	http.Handle("/metrics", promhttp.Handler())

	level.Info(logger).Log("msg", "Listening on address", "address", *listenAddress)
	if err := http.ListenAndServe(*listenAddress, nil); err != nil {
		level.Error(logger).Log("msg", "Error starting HTTP server", "err", err)
		os.Exit(1)
	}
}


collector.go
// +build linux

// ebpf_exporter - A Prometheus exporter for Linux block IO statistics.
//
// Copyright 2018 Daniel Swarbrick
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"bytes"
	"encoding/binary"
	"math"

	"github.com/iovisor/gobpf/bcc"
	"github.com/prometheus/client_golang/prometheus"
)

const (
	latTableLen   = 28 // Must match max_io_lat_slot in BPF program.
	reqSzTableLen = 16 // Must match max_io_req_sz_slot in BPF program.

	// cf. <linux/genhd.h>
	DISK_NAME_LEN = 32

	// Linux req_opf enums, cf. <linux/blk_types.h>
	REQ_OP_READ         = 0
	REQ_OP_WRITE        = 1
	REQ_OP_FLUSH        = 2
	REQ_OP_DISCARD      = 3
	REQ_OP_ZONE_REPORT  = 4
	REQ_OP_SECURE_ERASE = 5
	REQ_OP_ZONE_RESET   = 6
	REQ_OP_WRITE_SAME   = 7
	REQ_OP_WRITE_ZEROES = 9
	REQ_OP_SCSI_IN      = 32
	REQ_OP_SCSI_OUT     = 33
	REQ_OP_DRV_IN       = 34
	REQ_OP_DRV_OUT      = 35
)

var (
	// Map of request operation enums to human-readable strings. This map does not include all
	// possible request operations, but covers the most commonly observed ones.
	reqOpStrings = map[uint8]string{
		REQ_OP_READ:         "read",
		REQ_OP_WRITE:        "write",
		REQ_OP_FLUSH:        "flush",
		REQ_OP_DISCARD:      "discard",
		REQ_OP_WRITE_SAME:   "write_same",
		REQ_OP_WRITE_ZEROES: "write_zeroes",
	}
)

type exporter struct {
	bpfMod       *bcc.Module
	ioLat        *bcc.Table
	ioReqSz      *bcc.Table
	latency      *prometheus.Desc
	reqSize      *prometheus.Desc
	tableEntries *prometheus.GaugeVec
}

func newExporter(m *bcc.Module) *exporter {
	e := exporter{
		bpfMod:  m,
		ioLat:   bcc.NewTable(m.TableId("io_lat"), m),
		ioReqSz: bcc.NewTable(m.TableId("io_req_sz"), m),
		latency: prometheus.NewDesc(
			prometheus.BuildFQName(namespace, "bio", "req_latency"),
			"A histogram of bio request latencies in microseconds.",
			[]string{"device", "operation"},
			nil,
		),
		reqSize: prometheus.NewDesc(
			prometheus.BuildFQName(namespace, "bio", "req_size"),
			"A histogram of bio request sizes in KiB.",
			[]string{"device", "operation"},
			nil,
		),
		tableEntries: prometheus.NewGaugeVec(
			prometheus.GaugeOpts{
				Namespace: namespace,
				Subsystem: "bio",
				Name:      "bpf_table_entries",
				Help:      "The number of BPF table entries used.",
			},
			[]string{"table"},
		),
	}

	prometheus.MustRegister(e.tableEntries)

	return &e
}

func (e *exporter) Collect(ch chan<- prometheus.Metric) {
	var (
		n   uint
		tbl map[string]map[uint8][]uint64
	)

	n, tbl = decodeTable(e.ioLat, latTableLen)
	e.tableEntries.WithLabelValues("req_latency").Set(float64(n))
	e.emit(ch, e.latency, tbl)

	n, tbl = decodeTable(e.ioReqSz, reqSzTableLen)
	e.tableEntries.WithLabelValues("req_size").Set(float64(n))
	e.emit(ch, e.reqSize, tbl)
}

func (e *exporter) Describe(ch chan<- *prometheus.Desc) {
	ch <- e.latency
	ch <- e.reqSize
}

func (e *exporter) emit(ch chan<- prometheus.Metric, hist *prometheus.Desc, devBuckets map[string]map[uint8][]uint64) {
	for devName, reqs := range devBuckets {
		for reqOp, bpfBuckets := range reqs {
			var (
				count uint64
				sum   float64
			)

			// Skip unrecognized request operations.
			if _, ok := reqOpStrings[reqOp]; !ok {
				continue
			}

			promBuckets := make(map[float64]uint64)

			for k, v := range bpfBuckets {
				// Prometheus histograms are cumulative, so count must be a running total of
				// previous buckets also.
				count += v

				// Sum will not be completely accurate, since the BPF program already discarded
				// some resolution when storing occurrences of values in log2 buckets. Count and
				// sum are required however to calculate an average from a histogram.
				exp2 := math.Exp2(float64(k))
				sum += exp2 * float64(v)

				promBuckets[exp2] = count
			}

			ch <- prometheus.MustNewConstHistogram(hist,
				count,
				sum,
				promBuckets,
				devName, reqOpStrings[reqOp],
			)
		}
	}
}

// decodeTable decodes a BPF table and returns a per-device map of values as ordered buckets.
func decodeTable(table *bcc.Table, tableSize uint) (uint, map[string]map[uint8][]uint64) {
	var (
		numEntries uint

		// Struct must match disk_key_t in BPF program
		key struct {
			Disk [DISK_NAME_LEN]byte
			Op   uint8
			_    [7]byte // padding
			Slot uint64
		}
	)

	devBuckets := make(map[string]map[uint8][]uint64)

	// bcc.Table.Iter() returns unsorted entries, so write the decoded values into an order-
	// preserving slice.
	for it := table.Iter(); it.Next(); {
		if err := binary.Read(bytes.NewReader(it.Key()), binary.LittleEndian, &key); err != nil {
			continue
		}

		// key.Disk is a null-terminated char array.
		devName := string(key.Disk[:bytes.IndexByte(key.Disk[:], 0)])

		// First time seeing this device, create map for request operations.
		if _, ok := devBuckets[devName]; !ok {
			devBuckets[devName] = make(map[uint8][]uint64)
		}

		// First time seeing this req op for this device, create slice for buckets.
		if _, ok := devBuckets[devName][key.Op]; !ok {
			devBuckets[devName][key.Op] = make([]uint64, tableSize)
		}

		devBuckets[devName][key.Op][key.Slot] = binary.LittleEndian.Uint64(it.Leaf())
		numEntries++
	}

	return numEntries, devBuckets
}
